helmtemplate/helm.go

```diff
@@ -25,24 +26,41 @@ const (
 type TemplateHelmChartTask struct {
 	fs afero.Fs
 
-	chartDir string
-	branch   string
-	sha      string
-	version  string
+	chartDir     string
+	branch       string
+	sha          string
+	chartVersion string
+	appVersion   string
 }
 
 // Config holds configuration for building a new TemplateHelmChartTask
 type Config struct {
 	Fs afero.Fs
 
-	ChartDir string
-	Branch   string
-	Sha      string
-	Version  string
+	ChartDir   string
+	Branch     string
+	Sha        string
+	Version    string
+	AppVersion string
+}
+
+// renderedChart is used for chart validation after it has been filled with
+// values
+type renderedChart struct {
+	Version    string `json:"version,omitempty"`
+	AppVersion string `json:"appVersion,omitempty"`
 }
 
 // Run templates the chart's Chart.yaml and templates/deployment.yaml.
-func (t TemplateHelmChartTask) Run() error {
+func (t TemplateHelmChartTask) Run(validate, taggedBuild bool) error {
+	// We expect versions to match for a tagged build if project.go has been found.
+	if validate && taggedBuild && t.appVersion != "" && t.chartVersion != t.appVersion {
+		return microerror.Maskf(
+			versionMismatchError,
```

Comment:
As fmt says. If matcher is not used then it should be `executionFailedError`.

```suggestion
			executionFailedError,
```